### ISSUE TITLE
Board Review: Management Plane Namespace Review <Azure.ResourceManager.Migrate>

### ISSUE THREAD

**Issue by kuchatur-ms** [Azure/azure-sdk#6964]
Thank you for submitting this review request. Thorough review of your management library namespaces ensures that your library names are consistent with the guidelines and the consumers of your management library have a consistently good experience when using Azure.

To ensure consistency, all language library names will generally be reviewed together.

**Before submitting, ensure you adjust the title of the issue appropriately.**

**Note that the required material must be included before a meeting can be scheduled.**

## Contacts and Timeline

* Responsible service team: Azure Migrate engineering team.
* Main contacts: kuchatur, sunising
* Related contacts: tarsharm
* Expected code complete date: 10-12-23
* Expected release date: 14-01-24

## About the Service (required)
namespace name: Microsoft.Migrate
* Link to documentation introducing/describing the service: https://learn.microsoft.com/en-us/azure/migrate/
* Link to the service REST APIs Specs: https://github.com/Azure/azure-rest-api-specs/tree/main/specification/migrate/resource-manager/Microsoft.Migrate
* Link to GitHub issue for previous review sessions, if applicable:  NA

## Namespace Proposals (required per language)

**In the examples below please replace every occurrence of [ResourceProviderName] with your the service resource provider name. Be sure to keep the casing shown with [ResourceProviderName] when replacing it with the service resource provider name.**

- **.NET:** `Azure.ResourceManager.Migrate`
- **Java:** `azure-resourcemanager-migrate (com.azure.resourcemanager.migrate)`
- **Go/Golang:** `sdk/resourcemanager/migrate/armmigrate`
- **JavaScript:** `@azure/arm-migrate`
- **Python:** `azure-mgmt-migrate`

## Thank you!

**Comment by ronniegeraghty** [comment]
Hi @kuchatur-ms,
Thank you for submitting a request for a MGMT Plane Namespace Review. 

On the namespaces you proposed, we generally don't use the Microsoft prefix from the RP. 

Please see my answers below on your questions.

> According to the issue template ive updated the package names. If you feel we can get more customized package names, do let us know. We will make the package names more user friendly. especially for Microsoft.OffAzure RP as that is not a customer friendly name (we will create a similar request for that RP soon)

We can use custom names for the packages. In most cases the RP just tends to be the best name. 
For this service my recommendation would be to go with `Azure.ResourceManager.Migrate`. @ArthurMa1978 please give your opinion on my suggestion. 
If you have a different suggestion, please comment it. 

> We have several services under Microsoft.Migrate RP and are planning to check-in seperate swaggers for them in the official repo. Can we (and should we) publish a common SDK or multiple SDKs according to the swaggers?

The answer to this question may depend on if you're strictly talking about MGMT libraries or Data Plane libraries. I'll ask @ArthurMa1978 to comment on the MGMT plane, but on the data plane side the answer would be yes. It usually better to break the library up if the API Surface of the service is large and has distinct parts. 

> We are trying to release the SDK by EOY, but I presume you will have release embargoes during year end. When can we release post that? Mid Jan?

I do believe there is a cut off date for release MGMT Plane libraries before the end of the year. @ArthurMa1978 can you please let us know what that date is. For Data plane libraries there is no in-band release cycle in December but teams can still release out-of-band, however I cannot guarantee the Azure SDK team will be able to support your team if you run into issues since a large amount of our team will be taking time off. We also generally recommend against releasing in December for this reason.

**Comment by kuchatur-ms** [comment]
> Hi @kuchatur-ms, Thank you for submitting a request for a MGMT Plane Namespace Review.
> 
> On the namespaces you proposed, we generally don't use the Microsoft prefix from the RP.
> 
> Please see my answers below on your questions.
> 
> > According to the issue template ive updated the package names. If you feel we can get more customized package names, do let us know. We will make the package names more user friendly. especially for Microsoft.OffAzure RP as that is not a customer friendly name (we will create a similar request for that RP soon)
> 
> We can use custom names for the packages. In most cases the RP just tends to be the best name. For this service my recommendation would be to go with `Azure.ResourceManager.Migrate`. @ArthurMa1978 please give your opinion on my suggestion. If you have a different suggestion, please comment it.
> 
> > We have several services under Microsoft.Migrate RP and are planning to check-in seperate swaggers for them in the official repo. Can we (and should we) publish a common SDK or multiple SDKs according to the swaggers?
> 
> The answer to this question may depend on if you're strictly talking about MGMT libraries or Data Plane libraries. I'll ask @ArthurMa1978 to comment on the MGMT plane, but on the data plane side the answer would be yes. It usually better to break the library up if the API Surface of the service is large and has distinct parts.
> 
> > We are trying to release the SDK by EOY, but I presume you will have release embargoes during year end. When can we release post that? Mid Jan?
> 
> I do believe there is a cut off date for release MGMT Plane libraries before the end of the year. @ArthurMa1978 can you please let us know what that date is. For Data plane libraries there is no in-band release cycle in December but teams can still release out-of-band, however I cannot guarantee the Azure SDK team will be able to support your team if you run into issues since a large amount of our team will be taking time off. We also generally recommend against releasing in December for this reason.

Thanks for your replies.
- Yes Azure.ResourceManager.Migrate looks good to me as a customized name. I'll confirm with PMs on this and get back to you.
- We are releasing SDKs for management plane APIs. These are owned by multiple services under single namespace. I'd like to understand if we can break up the SDK under the namespace like this
- - Azure.ResourceManager.Migrate.Assessment, Azure.ResourceManager.Migrate.Migration, Azure.ResourceManager.Migrate.Modernization etc
- Will wait for Arthur's responses.

**Comment by ronniegeraghty** [comment]
> We are releasing SDKs for management plane APIs. These are owned by multiple services under single namespace. I'd like to understand if we can break up the SDK under the namespace like this
Azure.ResourceManager.Migrate.Assessment, Azure.ResourceManager.Migrate.Migration, Azure.ResourceManager.Migrate.Modernization etc

Generally, we try to only 3 levels for our Mgmt Namespaces. 4 levels can be done but will need to be approved by our architects. For the names you suggested, they look okay except for `Azure.ResourceManager.Migrate.Migration`. That name seems very repetitive and a little confusing.

**Comment by kuchatur-ms** [comment]
Hi @ronniegeraghty ,
I have confirmation from PMs that we'd like to go ahead with Azure.ResourceManager.Microsoft.Migrate name
We are not splitting the SDK right now. Please review/approve

**Comment by ArthurMa1978** [comment]
Hey @kuchatur-ms,
As @ronniegeraghty said `We can use custom names for the packages. In most cases the RP just tends to be the best name`, such as the package namespace of the released SDKs for [Microsoft.Migrate](https://github.com/Azure/azure-rest-api-specs/tree/main/specification/resourcemover/resource-manager) is `ResourceMover`, not as the RP namespace.

We highly recommend release one SDK for one RP namespace, but acceptable if you have broken up the swagger into different folder. 

All Microsoft Azure services share the common prefix `Azure.ResourceManager`. To avoid redundancy, the third level of the namespace should not contain `Microsoft` or `Azure`, even if they are part of the RP namespace.
The namespace should end with a noun that describes the service, not a verb. Based on the service description, my suggestion is:
- .NET: `Azure.ResourceManager.MigrateHub`
- Java: `azure-resourcemanager-migratehub (com.azure.resourcemanager.migratehub)`
- Go/Golang: `sdk/resourcemanager/migratehub/migratehub`
- JavaScript: `@azure/arm-migratehub`
- Python: `azure-mgmt-migratehub`

Release SDKs for all 5 languages BOY is too tight. You need to get the namespace approved before you can proceed with the release. For Java, Python, JS and Go, you can file the [tickets here](https://portal.azure-devex-tools.com/tools/request-api-release). For .Net, you can generate the SDK by yourself following this [guideline](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/78/Mgmt-.NET-SDK-release-process). I think next January is a reasonable timeline for us.

**Comment by ronniegeraghty** [comment]
@ArthurMa1978 and @kuchatur-ms, 
Is this a namespace change for an already released library? The Resource Mover libraries were already released. Are these new libraries a replacement to the Resource Mover libraries or do the new libraries map to a different API Spec?

**Comment by kuchatur-ms** [comment]
@ArthurMa1978 thanks for sharing the link to SDK generation for .net
I have a couple of questions regarding the process and need to get an idea of how long does this process usually take

1. According to your experience, how long does the E2E process of generating a new SDK and releasing it to public take? Just looking for a rough estimate from other teams' efforts.
2. While we have published our swagger in rest-api-specs repo, is there a way to generate SDK while we are developing the swagger, so that we can test the APIs with an SDK while the feature is in development? Something like a PR build when the PR is pending merge?
3. We are looking to follow a RESTAPI first process going forward and looks like the process is to get the API contracts merged into the https://github.com/azure/azure-sdk-for-net-pr repo early in the development phase. So do we use the SDK generated from this repo to work on our E2E tests etc before the feature is released? If you have a suggested guideline, it would be helpful.

**Comment by maririos** [comment]
Hey @kuchatur-ms curious, do you have a [Release Plan created](https://eng.ms/docs/products/azure-developer-experience/plan/release-plan) for your work? It guides you on the process from API Specs to SDK release and can give you highlight of the process and the steps you need to do to get the SDKs out the door.
You can also schedule an [informational meeting](https://eng.ms/docs/products/azure-developer-experience/onboard#schedule-an-informational-meeting) with our PM team to get clarifying questions about your timeline and commitments.

**Comment by ArthurMa1978** [comment]
@kuchatur-ms 
For your questions:
1. No, there is no single package that includes multiple SDKs. This is why we advise you to avoid splitting your SDKs into separate packages if they have common components and dependencies.
2. The .Net SDK can be completed in two weeks once the namespace is approved. For other languages, the release cycle is monthly.
3. You can follow the [guideline](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/78/Mgmt-.NET-SDK-release-process) generate the sdk from local folder.
4. Similar to the third question, you can generate the SDK from any location where the swagger spec is stored.

In fact, other services usually release different resources in one SDK by grouping the swagger spec with a release tag: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/compute/resource-manager/readme.md?plain=1#L252.
Depending on your needs, you can either combine them into one package or keep them as separate packages.

**Comment by kuchatur-ms** [comment]
Hi @maririos . Yes we have a release planner we are using to track this sdk release. https://apps.powerapps.com/play/e/ed2ffefd-774d-40dd-ab23-7fff01aeec9f/a/821ab569-ae60-420d-8264-d7b5d5ef734c?release-plan-id=35d00d60-876e-ee11-8df0-000d3a340b2c&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47
@ladonnaq is helping us with the SDK release.

@ArthurMa1978 , thanks for your insights.
We have 1 RP (Microsoft.Migrate) which contains 3-4 tracked resources
Each tracked resource has multiple APIs exposed via proxy resources and these APIs are owned by different services. If i understand correctly, your suggestion is to maintain seperate swagger files and to add them to the same SDK package tag. This looks ok to me overall. Can you help me with these questions before we decide on single vs multiple SDKs
1. Does this reduce the independence of various sister teams working on independant features and API versions when releasing their version of the same SDK?
2. Can we see issues where there are naming conflicts of enums or classes of the same name? As the multiple swaggers will all be for the same RP

If there are no issues with using the same SDK, then i dont see much reason in seperating the SDKs as that would be painful for the end users.

**Comment by ArthurMa1978** [comment]
Hey @kuchatur-ms, for your questions:
1. Each team works on a different feature, and they can either be independent or share common components.
2. Yes, this needs manual intervention to prevent it.
If you agree on releasing one package, please start with the swagger, which is now in a separate folder.

**Comment by kuchatur-ms** [comment]
Hi, discussed internally and we are good with 1 SDK for migrate RP. Here are the final name suggestions from our end.

.NET: Azure.ResourceManager.Migrate
Java: azure-resourcemanager-migrate (com.azure.resourcemanager.migrate)
Go/Golang: sdk/resourcemanager/Microsoft.Migrate/armmigrate
JavaScript: @azure/arm-migrate
Python: azure-mgmt-migrate

- Our swaggers for the SDK we want to publish are in the migrate/ folder in rest-api-spec repo already.
- We want to release an SDK for the APEX PLR work done recently, ie 2023-03-15 API version. We have all swaggers ready for that.
- Resource mover team will move their swagger into our repo soon, and then we will merge the already released resourcemover SDK into the new SDK we want to release in future.

@ArthurMa1978 , @ronniegeraghty lemme know if we are good to go from your end.

**Comment by ronniegeraghty** [comment]
@kuchatur-ms, fixing the Go namespace: 
.NET: Azure.ResourceManager.Migrate
Java: azure-resourcemanager-migrate (com.azure.resourcemanager.migrate)
Go/Golang: sdk/resourcemanager/migrate/armmigrate
JavaScript: @azure/arm-migrate
Python: azure-mgmt-migrate

I've updated the main description of this issues with these new names. 
Let me know if the Go name still looks okay. If so, we can show the names to the rest of the architects and see if there are any objections. I don't foresee there being any, but we do have to give them a week to make any objections.

**Comment by kuchatur-ms** [comment]
Yes this looks good. We can start with the architect review.

**Comment by ArthurMa1978** [comment]
@kuchatur-ms please don't forget to create a new tag in the [readme file](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/migrate/resource-manager/readme.md) which include both Microsoft.Migrate and Microsoft.OffAzure

**Comment by kuchatur-ms** [comment]
Hi.
For now we are releasing 1 SDK per RP so the migrate SDK is for microsoft.migrate only.
OffAzure will be released as a seperate SDK.

**Comment by kuchatur-ms** [comment]
@ArthurMa1978 , @ronniegeraghty is there anything else required from my end? If not, can you help complete the review and close the issue?

**Comment by kuchatur-ms** [comment]
Hi @ronniegeraghty , @ArthurMa1978 . Ping on this..
Can we close this now?

**Comment by ronniegeraghty** [comment]
Hi @kuchatur-ms, 
Sounds like we have agreed that the names in the main description of this issue are good for the ARM APIs under Microsoft.Migrate.

The next step is to present the names to the rest of our architect team. They'll have 1 week to make any objections. If there are no objections after the week then the names will be considered approved, and I'll update this issue saying so.

**Comment by ronniegeraghty** [comment]
The week is up and there were no objections from our architects on the proposed names. The names in the description of this issue are considered approved.
